DigitalLearningSolutions.Data/Services/CourseDelegatesDownloadFileService.cs

```diff
@@ -0,0 +1,177 @@
+            ICourseService courseService,
+            ICourseDelegatesDataService courseDelegatesDataService,
+            ICourseAdminFieldsService courseAdminFieldsService,
+            ICentreCustomPromptsService customPromptsService
+        )
+        {
+            this.courseService = courseService;
+            this.courseDelegatesDataService = courseDelegatesDataService;
+            this.courseAdminFieldsService = courseAdminFieldsService;
+            this.customPromptsService = customPromptsService;
+        }
+
+        public byte[] GetCourseDelegateDownloadFileForCourse(int customisationId, int centreId, int? adminCategoryId)
+        {
+            using var workbook = new XLWorkbook();
+
+            PopulateCourseDelegatesSheetForCourse(workbook, customisationId, centreId, adminCategoryId);
+
+            using var stream = new MemoryStream();
+            workbook.SaveAs(stream);
+            return stream.ToArray();
+        }
+
+        private void PopulateCourseDelegatesSheetForCourse(
+            IXLWorkbook workbook,
+            int customisationId,
+            int centreId,
+            int? adminCategoryId
+        )
+        {
+            var course = courseService.GetCourseDetailsFilteredByCategory(customisationId, centreId, adminCategoryId)!;
+            var adminFields = courseAdminFieldsService.GetCustomPromptsForCourse(customisationId);
+
+            var customRegistrationPrompts = customPromptsService.GetCustomPromptsForCentreByCentreId(centreId);
+
+            var courseDelegates = courseDelegatesDataService.GetDelegatesOnCourseForExport(customisationId, centreId);
+
+            var dataTable = new DataTable();
+
+            SetUpDataTableColumns(customRegistrationPrompts, adminFields, dataTable);
+
+            foreach (var courseDelegate in courseDelegates.OrderBy(x => x.LastName))
+            {
+                AddDelegateToDataTable(dataTable, courseDelegate, customRegistrationPrompts, adminFields);
+            }
+
+            ClosedXmlHelper.AddSheetToWorkbook(
+                workbook,
+                $"Course {course.CustomisationId}",
+                dataTable.AsEnumerable(),
+                XLTableTheme.None
+            );
+        }
+
+        private static void SetUpDataTableColumns(
+            CentreCustomPrompts customRegistrationPrompts,
+            CourseAdminFields adminFields,
+            DataTable dataTable
+        )
+        {
+            dataTable.Columns.AddRange(
+                new[] { new DataColumn("Last name"), new DataColumn("First name"), new DataColumn("Email") }
+            );
+
+            foreach (var prompt in customRegistrationPrompts.CustomPrompts)
+            {
+                dataTable.Columns.Add(
+                    !dataTable.Columns.Contains(prompt.CustomPromptText)
+                        ? prompt.CustomPromptText
+                        : $"{prompt.CustomPromptText} (Prompt {prompt.CustomPromptNumber})"

```

Comment:
The datatable columns I'm setting up can't have the same name, but there is nothing to stop a centre setting up registration prompts or admin fields with the same prompt. This just handles that case while allowing the file to generate with a sensible column name. I would assume that they wouldn't set up these prompts that this would happen, but given that I know it can, it's probably better to just handle it now.